### PR TITLE
A0-2257: Track highest justified block in the forest

### PR DESCRIPTION
**Pull request by lesniak43**
# Description

Currently we don’t request blocks that are not below a top required block, and have imported justifications, therefore we won’t get older justifications.

This happens e.g. when we get broadcasted blocks, and import them before anyone tells us that the block is required.

We need to track current highest justified block, and spawn a task for it.

## Type of change

- Bug fix (non-breaking change which fixes an issue)

# Checklist:

- I have added tests
- I have made corresponding changes to the existing documentation
- I have created new documentation

### CODE REVIEW

File: finality-aleph/src/sync/forest/mod.rs [hunk 1]
```diff
@@ -392,11 +413,18 @@ impl<I: PeerId, J: Justification> Forest<I, J> {
         use VertexHandle::Candidate;
         match self.get_mut(id) {
             Candidate(entry) => {
-                // request only required blocks
+                let know_most = entry.get().vertex.know_most().clone();
+                // request only required blocks, or the highest_justified block/header
                 if !&entry.get().vertex.required() {
-                    return None;
+                    match &self.highest_justified {
```
Comment (timorl) [inline finality-aleph/src/sync/forest/mod.rs:419]: `if Some(id) != self.highest_justified`

File: finality-aleph/src/sync/forest/mod.rs [hunk 2]
```diff
@@ -409,16 +437,26 @@ impl<I: PeerId, J: Justification> Forest<I, J> {
...
-                    know_most,
-                    branch_knowledge,
-                },
-                false => Interest::Required {
-                    know_most,
-                    branch_knowledge,
-                },
-            },
+            Some((know_most, branch_knowledge)) => {
+                if let Some(highest_justified_id) = &self.highest_justified {
```
Comment (timorl) [inline finality-aleph/src/sync/forest/mod.rs:441]: Same here

File: finality-aleph/src/sync/forest/mod.rs (hunk 1 above)
Comment (lesniak43) [reply finality-aleph/src/sync/forest/mod.rs]: fixed

File: finality-aleph/src/sync/forest/mod.rs (hunk 2 above)
Comment (lesniak43) [reply finality-aleph/src/sync/forest/mod.rs]: fixed

File: finality-aleph/src/sync/handler.rs [hunk 3]
```diff
@@ -206,6 +209,19 @@ impl<I: PeerId, J: Justification, CS: ChainStatus<J>, V: Verifier<J>, F: Finaliz
             false => None,
         };
         self.try_finalize()?;
+        if let Some(id) = &maybe_id {
```
Comment (timorl) [inline finality-aleph/src/sync/handler.rs:212]: This isn't really necessary, as we check the state after taking things out from the task queue anyway.

The comment above should also then be changed to "Returns an ID if this justification was higher than any previously known." or something similar.

File: finality-aleph/src/sync/handler.rs [hunk 4]
```diff
@@ -257,6 +273,8 @@ impl<I: PeerId, J: Justification, CS: ChainStatus<J>, V: Verifier<J>, F: Finaliz
     }
 
     /// Handle a single justification.
+    /// Return `Some(id)` if the highest known justification has just changed to
+    /// the one identified by `id`, unless it's also just been finalized.
```
Comment (timorl) [inline finality-aleph/src/sync/handler.rs:277]: If you change as suggested above this comment also requires a similar change.

File: finality-aleph/src/sync/service.rs [hunk 5]
```diff
@@ -164,18 +164,26 @@ impl<
...
         let mut previous_block_id = None;
         for justification in justifications {
             let maybe_block_id = match self
                 .handler
-                .handle_justification(justification, Some(peer.clone()))
+                .handle_justification(justification, peer.clone())
             {
                 Ok(maybe_id) => maybe_id,
                 Err(e) => {
+                    let peer_id = match peer {
```
Comment (timorl) [inline finality-aleph/src/sync/service.rs:180]: Consider using `map_or`.

File: finality-aleph/src/sync/service.rs [hunk 6]
```diff
@@ -213,21 +221,30 @@ impl<
...
         }
     }
 
     fn handle_task(&mut self, block_id: BlockIdFor<J>) {
         use Interest::*;
         match self.handler.block_state(&block_id) {
+            HighestJustified {
+                know_most,
+                branch_knowledge,
+            } => {
```
Comment (timorl) [inline finality-aleph/src/sync/service.rs:244]: I believe this could be merged with the case below using `||`.

**Review by timorl — APPROVED** [review]
Looks nice, only a couple of nits.

File: finality-aleph/src/sync/forest/mod.rs [hunk 7]
```diff
@@ -347,6 +362,12 @@ impl<I: PeerId, J: Justification> Forest<I, J> {
                 }
             }
         }
+        // if root is the only justified block now, set cache to None
```
[the remark this answers is not in the document]
Comment (lesniak43) [reply finality-aleph/src/sync/forest/mod.rs:365]: fixed, also fixed other comments in this method

File: finality-aleph/src/sync/forest/mod.rs [hunk 8]
```diff
@@ -285,35 +291,44 @@ impl<I: PeerId, J: Justification> Forest<I, J> {
...
     pub fn update_justification(
         &mut self,
         justification: J,
         holder: Option<I>,
     ) -> Result<bool, Error> {
-        use JustificationAddResult::*;
         let (id, parent_id) = self.process_header(justification.header())?;
-        self.update_header(justification.header(), None, false)?;
-        match self.get_mut(&id) {
+        // set required
```
[the remark this answers is not in the document]
Comment (lesniak43) [reply finality-aleph/src/sync/forest/mod.rs:319]: removed

File: finality-aleph/src/sync/forest/mod.rs [hunk 9]
```diff
@@ -392,11 +413,11 @@ impl<I: PeerId, J: Justification> Forest<I, J> {
         use VertexHandle::Candidate;
         match self.get_mut(id) {
             Candidate(entry) => {
-                // request only required blocks
-                if !&entry.get().vertex.required() {
+                let know_most = entry.get().vertex.know_most().clone();
```
Comment (kostekIV) [inline finality-aleph/src/sync/forest/mod.rs:416]: nit: previous order looked better - it avoided unnecessary clone

File: finality-aleph/src/sync/handler.rs (hunk 3 above)
Comment (lesniak43) [reply finality-aleph/src/sync/handler.rs:212]: fixed

File: finality-aleph/src/sync/handler.rs (hunk 4 above)
Comment (lesniak43) [reply finality-aleph/src/sync/handler.rs:277]: fixed

**Review by kostekIV — APPROVED** [review]

File: finality-aleph/src/sync/service.rs (hunk 5 above)
Comment (lesniak43) [reply finality-aleph/src/sync/service.rs:180]: fixed

File: finality-aleph/src/sync/forest/mod.rs [hunk 10]
```diff
@@ -347,6 +362,12 @@ impl<I: PeerId, J: Justification> Forest<I, J> {
                 }
             }
         }
+        // if root is the only justified block now, set cache to None
+        if let Some(id) = &self.highest_justified {
+            if id.number() <= self.root_id.number() {
```
[the remark this answers is not in the document]
Comment (lesniak43) [reply finality-aleph/src/sync/forest/mod.rs]: fixed

**Pull request merged**